src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,111 +36,316 @@
     ZIP_FILE_EXTENSIONS + TAR_FILE_EXTENSIONS + LZMA_FILE_EXTENSIONS)
 
 
-class ArchiveType:
-  """Type of the archive."""
-  UNKNOWN = 0
-  ZIP = 1
-  TAR = 2
-  TAR_LZMA = 3
+def _is_attempting_path_traversal(archive_name, output_dir, filename) -> bool:
+  """Detects whether there is a path traversal attempt.
 
+  Args:
+      archive_name (str): the name of the archive.
+      output_dir (str): the output directory.
+      filename (str): the name of the file being checked.
 
-class ArchiveFile:
-  """File in an archive."""
+  Returns:
+      bool: Whether there is a path traversal attempt
+  """
+  output_dir = os.path.realpath(output_dir)
+  absolute_file_path = os.path.join(output_dir, os.path.normpath(filename))
+  real_file_path = os.path.realpath(absolute_file_path)
 
-  def __init__(self, name, size, handle):
-    self.name = name
-    self.size = size
-    self.handle = handle
+  if real_file_path == output_dir:
+    # Workaround for https://bugs.python.org/issue28488.
+    # Ignore directories named '.'.
+    return False
 
+  if real_file_path != absolute_file_path:
+    logs.log_error('Directory traversal attempted while unpacking archive %s '
+                   '(file path=%s, actual file path=%s). Aborting.' %
+                   (archive_name, absolute_file_path, real_file_path))
+    return True
+  return False
 
-def iterator(archive_path,
-             archive_obj=None,
-             file_match_callback=None,
-             should_extract=True):
-  """Return an iterator for files in an archive. Extracts files if
-  |should_extract| is True."""
-  archive_type = get_archive_type(archive_path)
 
-  if not file_match_callback:
+@dataclasses.dataclass
+class ArchiveMemberInfo:
+  """Represents an archive member.
+  """
+  filename: str
+  is_dir: bool
+  file_size_bytes: int
+  mode: int
+
 
-    def file_match_callback_fallback(_):
-      return True
+class ArchiveReader(abc.ABC):
+  """Abstract class for representing an archive reader. Abstract methods must
+  be overriden.
+  """
 
-    file_match_callback = file_match_callback_fallback
+  @abc.abstractmethod
+  def list_files(self) -> typing.List[ArchiveMemberInfo]:
+    """Lists all members contained in the archives.
 
-  def maybe_extract(extract_func, info):
-    """Returns an extracted file or None if it is not supposed to be extracted.
+    Returns:
+        typing.List[ArchiveMemberInfo]: All the archive members
     """
-    if should_extract:
-      return extract_func(info)
-    return None
+    raise NotImplementedError
 
-  if archive_type == ArchiveType.ZIP:
-    try:
-      with zipfile.ZipFile(archive_obj or archive_path) as zip_file:
-        for info in zip_file.infolist():
-          if not file_match_callback(info.filename):
-            continue
+  @abc.abstractmethod
+  def extract(self, member, path, trusted=False) -> str:
+    """Extracts `member` out of the archive to the provided path.
```

Comment:
What happens if the member is a directory? I assume contents are extracted too, can we document this?

Reply:
No, we only extract the directory, which means the directory will be created. I'll add a comment for this.

Reply:
Thanks!